lib/doorkeeper/oauth/pre_authorization.rb

```diff
@@ -7,17 +7,21 @@ class PreAuthorization
       validate :client, error: :invalid_client
       validate :scopes, error: :invalid_scope
       validate :redirect_uri, error: :invalid_redirect_uri
+      validate :code_challenge_method, error: :invalid_code_challenge_method
 
-      attr_accessor :server, :client, :response_type, :redirect_uri, :state
+
+      attr_accessor :server, :client, :response_type, :redirect_uri, :state, :code_challenge, :code_challenge_method
       attr_writer   :scope
 
       def initialize(server, client, attrs = {})
-        @server        = server
-        @client        = client
-        @response_type = attrs[:response_type]
-        @redirect_uri  = attrs[:redirect_uri]
-        @scope         = attrs[:scope]
-        @state         = attrs[:state]
+        @server                = server
+        @client                = client
+        @response_type         = attrs[:response_type]
+        @redirect_uri          = attrs[:redirect_uri]
+        @scope                 = attrs[:scope]
+        @state                 = attrs[:state]
+        @code_challenge        = attrs[:code_challenge]
+        @code_challenge_method = attrs[:code_challenge_method] || 'plain'
```

Comment:
I don't think it should ever fallback. I would suggest removing `|| 'plain'` entirely. The request should be extremely explicit especially if it's `plain`, since this is strongly advised against the the RFC.